src/3d/chunks/qgschunkedentity_p.cpp

```diff
@@ -0,0 +1,379 @@
+  {
+    QgsChunkListEntry *entry = mReplacementQueue->takeLast();
+    entry->chunk->unloadChunk();  // also deletes the entry
+    ++unloaded;
+  }
+
+  if ( mBboxesEntity )
+  {
+    QList<QgsAABB> bboxes;
+    Q_FOREACH ( QgsChunkNode *n, mActiveNodes )
+      bboxes << n->bbox();
+    mBboxesEntity->setBoxes( bboxes );
+  }
+
+  // start a job from queue if there is anything waiting
+  if ( !mActiveJob )
+    startJob();
+
+  mNeedsUpdate = false;  // just updated
+
+  qDebug() << "update: active " << mActiveNodes.count() << " enabled " << enabled << " disabled " << disabled << " | culled " << mFrustumCulled << " | loading " << mChunkLoaderQueue->count() << " loaded " << mReplacementQueue->count() << " | unloaded " << unloaded << " elapsed " << t.elapsed() << "ms";
+}
+
+void QgsChunkedEntity::setShowBoundingBoxes( bool enabled )
+{
+  if ( ( enabled && mBboxesEntity ) || ( !enabled && !mBboxesEntity ) )
+    return;
+
+  if ( enabled )
+  {
+    mBboxesEntity = new QgsChunkBoundsEntity( this );
+  }
+  else
+  {
+    mBboxesEntity->deleteLater();
+    mBboxesEntity = nullptr;
+  }
+}
+
+void QgsChunkedEntity::updateNodes( const QList<QgsChunkNode *> &nodes, QgsChunkQueueJobFactory *updateJobFactory )
+{
+  Q_FOREACH ( QgsChunkNode *node, nodes )
+  {
+    if ( node->state() == QgsChunkNode::QueuedForUpdate )
+    {
+      mChunkLoaderQueue->takeEntry( node->loaderQueueEntry() );
+      node->cancelQueuedForUpdate();
+    }
+    else if ( node->state() == QgsChunkNode::Updating )
+    {
+      cancelActiveJob();  // we have currently just one active job so that must be it
+    }
+
+    Q_ASSERT( node->state() == QgsChunkNode::Loaded );
+
+    QgsChunkListEntry *entry = new QgsChunkListEntry( node );
+    node->setQueuedForUpdate( entry, updateJobFactory );
+    mChunkLoaderQueue->insertLast( entry );
+  }
+
+  // trigger update
+  if ( !mActiveJob )
+    startJob();
+}
+
+
+void QgsChunkedEntity::update( QgsChunkNode *node, const SceneState &state )
+{
+  if ( Qgs3DUtils::isCullable( node->bbox(), state.viewProjectionMatrix ) )
+  {
+    ++mFrustumCulled;
+    return;
+  }
+
+  node->ensureAllChildrenExist();
+
+  // make sure all nodes leading to children are always loaded
+  // so that zooming out does not create issues
+  requestResidency( node );
+
+  if ( !node->entity() )
+  {
+    // this happens initially when root node is not ready yet
+    qDebug() << "BOOM!";
+    return;
+  }
+
+  //qDebug() << node->x << "|" << node->y << "|" << node->z << "  " << tau << "  " << screenSpaceError(node, state);
+
+  if ( screenSpaceError( node, state ) <= mTau )
+  {
+    // acceptable error for the current chunk - let's render it
+
+    mActiveNodes << node;
+  }
+  else if ( node->allChildChunksResident( mCurrentTime ) )
+  {
+    // error is not acceptable and children are ready to be used - recursive descent
+
+    QgsChunkNode *const *children = node->children();
+    for ( int i = 0; i < 4; ++i )
```

Comment:
What do you think about adding some kind of `size()` method on QgsChunkNode?

Reply:
By design there are always four children - I like when it is more explicit in the code - a call to size() would make me wonder if there's a dynamic array of children...